htlcswitch/quiescer.go

```diff
@@ -0,0 +1,123 @@
+package htlcswitch
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// quiescer is a state machine that tracks progression through the quiescence
+// protocol.
+type quiescer struct {
+	// chanID marks what channel we are managing the state machine for. This
+	// is important because the quiescer is responsible for constructing the
+	// messages we send out and the ChannelID is a key field in that
+	// message.
+	chanID lnwire.ChannelID
+
+	// weOpened indicates whether we were the original opener of the
+	// channel. This is used to break ties when both sides of the channel
+	// send Stfu claiming to be the initiator.
+	weOpened bool
+
+	// localInit indicates whether our path through this state machine was
+	// initiated by our node. This can be true or false independently of
+	// remoteInit.
+	localInit bool
+
+	// remoteInit indicates whether we received Stfu from our peer where the
+	// message indicated that the remote node believes it was the initiator.
+	// This can be true or false independently of localInit.
+	remoteInit bool
+
+	// sent tracks whether or not we have emitted Stfu for sending.
+	sent bool
+
+	// received tracks whether or not we have received Stfu from our peer.
+	received bool
+}
+
+// recvStfu is called when we receive an Stfu message from the remote.
+func (q *quiescer) recvStfu(msg lnwire.Stfu) error {
+	if q.received {
+		return fmt.Errorf("stfu already received for channel %v",
+			q.chanID)
+	}
+
+	q.received = true
+	q.remoteInit = msg.Initiator
```

Comment:
If the message we received indicates that the sender is the initiator, then the remote party initiated the process. I do think all of this can be cleaned up with the [ChannelParty](https://github.com/lightningnetwork/lnd/pull/8755/commits/77eb59baa10aac83dbe000193f0b3ee58d28da7d) type though. I'm thinking maybe I cherry-pick that commit out of the next PR and move it ahead of everything.